app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,26 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def github
+    oauth 'github'
+  end
+
+  def google_oauth2
+    oauth 'google_oauth2'
+  end
+
+  def oauth(provider)
+    user = User.from_omniauth(request.env["omniauth.auth"])
```

Comment:
I suggest you get a user from the property in the way below. We can call user all across the `OmniauthCallbacksController`
```ruby
private

def user
  @user ||= Users::OmniAuth.new(request.env['omniauth.auth']).user
end
```
You may ask, what the `Users::OmniAuth`? Let's create it:
```ruby
app/services/users/omniauth.rb

class Users::Omniauth
  def initialize(auth)
    @auth = auth
  end

  def user
    find_user || create_user
  end

  private

  attr_reader :auth

  def find_user
    ...
  end

  def create_user
    ...
  end
end
```
Since you create a service we can get rid of `User.from_omniauth` method. And let's cover `Users::OmniAuth` with test.